lib/internal/util/parse_args/parse_args.js

```diff
@@ -0,0 +1,315 @@
+'use strict';
+
+const {
+  ArrayPrototypeForEach,
+  ArrayPrototypePushApply,
+  ArrayPrototypeShift,
+  ArrayPrototypeSlice,
+  ArrayPrototypePush,
+  ArrayPrototypeUnshiftApply,
+  ObjectPrototypeHasOwnProperty: ObjectHasOwn,
+  ObjectEntries,
+  StringPrototypeCharAt,
+  StringPrototypeIncludes,
+  StringPrototypeIndexOf,
+  StringPrototypeSlice,
+} = primordials;
+
+const {
+  validateArray,
+  validateObject,
+  validateString,
+  validateUnion,
+  validateBoolean,
+} = require('internal/validators');
+
+const {
+  findLongOptionForShort,
+  isLoneLongOption,
+  isLoneShortOption,
+  isLongOptionAndValue,
+  isOptionValue,
+  isOptionLikeValue,
+  isShortOptionAndValue,
+  isShortOptionGroup,
+  objectGetOwn,
+  optionsGetOwn,
+} = require('internal/util/parse_args/utils');
+
+const {
+  codes: {
+    ERR_INVALID_ARG_VALUE,
+    ERR_PARSE_ARGS_INVALID_OPTION_VALUE,
+    ERR_PARSE_ARGS_UNKNOWN_OPTION,
+    ERR_PARSE_ARGS_UNEXPECTED_POSITIONAL,
+  },
+} = require('internal/errors');
+
+function getMainArgs() {
+  // This function is a placeholder for proposed process.mainArgs.
+  // Work out where to slice process.argv for user supplied arguments.
+
+  // Electron is an interesting example, with work-arounds implemented in
+  // Commander and Yargs. Hopefully Electron would support process.mainArgs
+  // itself and render this work-around moot.
+  //
+  // In a bundled Electron app, the user CLI args directly
+  // follow executable. (No special processing required for unbundled.)
+  // 1) process.versions.electron is either set by electron, or undefined
+  //    see https://github.com/electron/electron/blob/master/docs/api/process.md#processversionselectron-readonly
+  // 2) process.defaultApp is undefined in a bundled Electron app, and set
+  //    in an unbundled Electron app
+  //    see https://github.com/electron/electron/blob/master/docs/api/process.md#processversionselectron-readonly
+  // (Not included in tests as hopefully temporary example.)
+  /* c8 ignore next 3 */
+  if (process.versions && process.versions.electron && !process.defaultApp) {
```

Comment:
```suggestion
  if (process.versions.electron && !process.defaultApp) {
```
isn't this conditional enough?

Reply:
Userland can `delete process.versions`. We could use optional chaining though.

Reply:
What if userland code does `delete global.process`? FYI, we do have looser checks like https://github.com/nodejs/node/blob/09c329de9f67334dff08bcd4ed0675ce4a49482f/lib/internal/assert/assertion_error.js#L350, so I think it's fine if we don't consider userland code deleting process (why would they even do that lol?).

Reply:
I don't see a reason not to guard with optional chaining, I don't think we should do much beyond this.

Reply:
> What if userland code does `delete global.process`?

Core wouldn't be impacted because it doesn't rely on the global, `process` is one of the arguments passed to the CJS function wrapper: https://github.com/nodejs/node/blob/7681fdfd7d167030b9d65df8403d8043b1df8127/src/node_native_module.cc#L175-L188

> FYI, we do have looser checks like

Then I guess we should fix those.

> I think it's fine if we don't consider userland code deleting process (why would they even do that lol?)

I don't think it's fine (why would we not protect our code from that? It's not like adding an optional chaining is a big deal anyway)